File: Swift/Algorithm/Algorithm/Datastructure/LinkedList.swift
//
//  LinkedList.swift
//  Algorithm
//
//  Created by amoyio on 2018/9/11.
//  Copyright © 2018年 amoyio. All rights reserved.
//

import Foundation

class LinkedList<NodeType: CustomDebugStringConvertible & Equatable>: CustomDebugStringConvertible {
    class Node<T: CustomDebugStringConvertible>: CustomDebugStringConvertible {
        var value: T?
        var next: Node<T>?
        
        init(value: T, next: Node?) {
            self.value = value
            self.next = next
        }
        
        init() {}
        
        convenience init(_ value: T) {
            self.init()
            self.value = value
        }
        
        var debugDescription: String {
            guard let valueString = self.value else { return "Empty" }
            return "[\(valueString.debugDescription)]->\(next == nil ? "None" : "")"
        }
    }
    
    var dummyHead: Node<NodeType>
    private(set) var size: Int = 0
    init() {
        self.dummyHead = LinkedList.Node<NodeType>()
        self.size = 0
    }
    
    func isEmpty() -> Bool {
        return size == 0
    }
    
    func add(index: Int, value: NodeType) -> Bool {
        guard index >= 0 && index <= size else {
            return false
        }
        var preNode: Node? = dummyHead
        for _ in 0..<index {
            preNode = preNode?.next
        }
        preNode?.next = Node(value: value, next: preNode?.next)
        size += 1
        return true
    }
    
    func addFirst(value: NodeType) -> Bool {
        return add(index: 0, value: value)
    }
    
    func addLast(value: NodeType) -> Bool {
        return add(index: self.size, value: value)
    }
    
    func object(at index: Int) -> Node<NodeType>? {
        guard index >= 0 && index < size else {
            return nil
        }
        var selectNode: Node? = dummyHead
        for _ in 0...index {
            selectNode = selectNode?.next
        }
        return selectNode
    }
    
    func getFirst() -> Node<NodeType>? {
        return object(at: 0)
    }
    
    func getLast() -> Node<NodeType>? {
        return object(at: size - 1)
    }
    
    func contains(value: NodeType) -> Bool {
        var selectedNode = self.dummyHead.next
        while let cNode = selectedNode {
            if cNode.value == value {
                return true
            }
            selectedNode = cNode.next
        }
        return false
    }
    
    func update(value: NodeType, index: Int) -> Bool {
        guard index >= 0 && index < size else {
            return false
        }
        var selectedNode: Node? = dummyHead
        for _ in 0...index {
            selectedNode = selectedNode?.next
        }
        selectedNode?.value = value
        return true
    }
    
    func remove(at index: Int) -> Node<NodeType>? {
        guard index >= 0 && index < size else {
            return nil
        }
        var preNode: Node? = dummyHead
        for _ in 0..<index {
            preNode = preNode?.next
        }
        let node = preNode?.next
        preNode?.next = node?.next
        node?.next = nil
        self.size -= 1
        return node
    }
    
    func removeFirst() -> Node<NodeType>? {
        return self.remove(at: 0)
    }
    
    func removeLast() -> Node<NodeType>? {
        return self.remove(at: self.size - 1)
    }
    
    func removeValue(value: NodeType) -> Bool {
        var selectNode: Node? = dummyHead
        while let preNode = selectNode, let node = preNode.next {
            if node.value! == value {
                preNode.next = node.next
                return true
            }
            selectNode = node
        }
        return false
    }
    
    var debugDescription: String {
        var selectNode: Node? = dummyHead
        var displayStr = "HEAD->"
        while selectNode?.next != nil {
            displayStr += "[\(String(describing: selectNode?.next!.value))]"
            selectNode = selectNode?.next
        }
        return displayStr
    }
}

File: Python/LinkedList.py
from typing import Optional


class LinkedList:
    class Node:

        def __init__(self, value, next_node):
            self.value = value
            self.next_node = next_node

        def __str__(self):
            return "[{}]->{}".format(self.value, ("None" if self.next_node is None else ""))

    def __init__(self):
        self.dummay_head = LinkedList.Node(None, None)
        self.size = 0

    def get_size(self) -> int:
        return self.size

    def is_empty(self) -> bool:
        return self.size == 0

    def add_first(self, value):
        self.add(0, value)

    def add_last(self, value):
        self.add(self.size, value)

    def add(self, index, value) -> bool:
        if not index >= 0 and index < self.size:
            return False
        pre_node = self.dummay_head
        for i in range(index):
            pre_node = pre_node.next_node
        pre_node.next_node = LinkedList.Node(value, pre_node.next_node)
        self.size += 1
        return True

    def object_at(self, index) -> Optional[Node]:
        if not index >= 0 and index < self.size:
            return None
        selected_node: LinkedList.Node = self.dummay_head.next_node
        for i in range(index):
            selected_node = selected_node.next_node
        return selected_node

    def get_first(self) -> Optional[Node]:
        return self.object_at(0)

    def get_last(self) -> Optional[Node]:
        return self.object_at(self.size - 1)

    def update_at(self, index, value) -> bool:
        if not index >= 0 and index < self.size:
            return False
        selected_node: LinkedList.Node = self.dummay_head.next_node
        for i in range(index):
            selected_node = selected_node.next_node
        selected_node.value = value
        return True

    def contains(self, value) -> bool:
        selected_node = self.dummay_head.next_node
        while selected_node != None:
            if selected_node.value == value:
                return True
            selected_node = selected_node.next_node
        return False

    def __str__(self):
        selected_node = self.dummay_head.next_node
        display_string = "HEAD->"
        while selected_node:
            display_string += selected_node.__str__()
            selected_node = selected_node.next_node
        return display_string

    def remove(self, index) -> Optional[Node]:
        if not index >= 0 and index < self.size:
            return None
        pre_node = self.dummay_head
        for item in range(index):
            pre_node = pre_node.next_node
        node = pre_node.next_node
        pre_node.next_node = node.next_node
        node.next_node = None
        self.size -= 1
        return node

    def remove_first(self) -> Optional[Node]:
        return self.remove(0)

    def remove_last(self) -> Optional[Node]:
        return self.remove(self.size - 1)

    def to_list(self) -> list:
        result = []
        selected_node = self.dummay_head.next_node
        while selected_node:
            result += selected_node.value
            selected_node = selected_node.next_node
        return result

